### PR TITLE
Implement kiai flash of hit circles

### LINKED ISSUE #7174: Kiai flash of hit circles is missing
**Describe the new feature:** In Kiai mode, hit circles in osu!standard would flash synced to beat. Feature already in osu!stable.

### PR DESCRIPTION
**Pull request by mcendu**
Closes #7174

### CODE REVIEW

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Osu/Skinning/LegacyMainCirclePiece.cs [hunk 1]
```diff
@@ -44,6 +45,19 @@ private void load(DrawableHitObject drawableObject, ISkinSource skin)
                     Anchor = Anchor.Centre,
                     Origin = Anchor.Centre,
                 },
+                new KiaiFlash
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Size = new Vector2(0.921875f),
```
[the remark this answers is not in the document]
Comment (mcendu) [reply osu.Game.Rulesets.Osu/Skinning/LegacyMainCirclePiece.cs]: Actual circle size (118) / Sprite size (128)

[the remark this answers is not in the document]
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Osu/Skinning/LegacyMainCirclePiece.cs]: Should put `118f / 128f` and add an inline comment referencing the circle and sprite sizes, putting the result with no reference is not accepted I believe.

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Skinning/LegacyMainCirclePiece.cs]: `Size = OsuLegacySkinTransformer. LEGACY_CIRCLE_RADIUS * 2` should be enough.

[the remark this answers is not in the document]
Comment (mcendu) [reply osu.Game.Rulesets.Osu/Skinning/LegacyMainCirclePiece.cs]: > `Size = OsuLegacySkinTransformer. LEGACY_CIRCLE_RADIUS * 2` should be enough.

Should be divided by `128` to make it fit the circle.

File: osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/KiaiFlash.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Shapes;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Containers;
+
+namespace osu.Game.Rulesets.Osu.Objects.Drawables.Pieces
+{
+    public class KiaiFlash : BeatSyncedContainer
+    {
+        public Drawable FlashComponent { get; set; }
```
[the remark this answers is not in the document]
Comment (mcendu) [reply osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/KiaiFlash.cs]: To ease implementation of flashing glow behind circles, but later thought that `Child` also fulfills that task so removed this property.

File: osu.Game.Rulesets.Osu/Skinning/LegacyMainCirclePiece.cs [hunk 3]
```diff
@@ -44,6 +46,25 @@ private void load(DrawableHitObject drawableObject, ISkinSource skin)
...
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    Size = new Vector2(OsuLegacySkinTransformer.LEGACY_CIRCLE_RADIUS / 64),
+
+                    Masking = true,
+                    CornerRadius = Size.X / 2,
+                    CornerExponent = 2,
+
+                    Intensity = 0.1f,
+                    Child = new Box
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/LegacyMainCirclePiece.cs]: I'm confused as to why this child box isn't in the `KiaiFlash` class itself. Also not sure a legacy implementation should be made with a separate design unless it matches stable 1:1 (this code doesn't look to be).

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
How are you testing this? Doesn't look like any tests were added.

**Comment by peppy** [comment]
@mcendu any updates? This still needs a basic test method added int he osu! ruleset tests.

**Comment by peppy** [comment]
I've tidied this up and added a legacy implementation.

Couple of notes:
- Doesn't include the slider body, which stable does, but this PR is focused on hitcircles so I think that's fine for now.
- The test doesn't seem to correctly actuate kiai flashes if run at <100% playback speed, not sure why. Needs further investigation but probably fine for now.
- To keep things simple I haven't brought across one of the stable timing details which is that if a kiai flash is going to exceed the `EndTime` of a hitobject it will be faded out faster to meet that point. I don't think this is required (was probably a workaround for animations not playing out correctly if the kiai one overlapped hit explode animations).

**Review by peppy — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**